rust/src/rfb/rfb.rs

```diff
@@ -633,3 +633,86 @@ pub unsafe extern "C" fn rs_rfb_register_parser() {
         SCLogDebug!("Protocol detector and parser disabled for RFB.");
     }
 }
+
+#[cfg(test)]
+mod test {
+    use super::*;
+
+    #[test]
+    fn test_error_state() {
+        let mut state = RFBState::new();
+
+        let buf = &[
+            0x05, 0x00, 0x03, 0x20, 0x20, 0x18, 0x00, 0x01, 0x00, 0xff, 0x00, 0xff, 0x00, 0xff,
+            0x10, 0x08, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x1e, 0x61, 0x6e, 0x65, 0x61,
+            0x67, 0x6c, 0x65, 0x73, 0x40, 0x6c, 0x6f, 0x63, 0x61, 0x6c, 0x68, 0x6f, 0x73, 0x74,
+            0x2e, 0x6c, 0x6f, 0x63, 0x61, 0x6c, 0x64, 0x6f, 0x6d, 0x61, 0x69, 0x6e,
+        ];
+        let r = state.parse_request(buf);
+
+        assert_eq!(
+            r,
+            AppLayerResult {
+                status: -1,
+                consumed: 0,
+                needed: 0
+            }
+        );
+    }
+    #[test]
+    fn test_request_response_events() {
+        let mut init_state = RFBState::new();
+
+        let buf: &[u8] = &[
+            0x52, 0x46, 0x42, 0x20, 0x30, 0x30, 0x33, 0x2e, 0x30, 0x30, 0x38, 0x0a,
+            0x01, /* Number of security types: 1 */
+            0x02, /* Security type: VNC (2) */
+            0x02, /* Security type selected: VNC (2) */
+            0x54, 0x7b, 0x7a, 0x6f, 0x36, 0xa1, 0x54, 0xdb, 0x03, 0xa2, 0x57, 0x5c, 0x6f, 0x2a,
+            0x4e,
+            0xc5, /* 16 byte Authentication challenge: 547b7a6f36a154db03a2575c6f2a4ec5 */
+            0x00, 0x00, 0x00, 0x00, /* Authentication result: OK */
+            0x00, /* Share desktop flag: False */
+            0x05, 0x00, 0x03, 0x20, 0x20, 0x18, 0x00, 0x01, 0x00, 0xff, 0x00, 0xff, 0x00, 0xff,
+            0x10, 0x08, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x1e, 0x61, 0x6e, 0x65, 0x61,
+            0x67, 0x6c, 0x65, 0x73, 0x40, 0x6c, 0x6f, 0x63, 0x61, 0x6c, 0x68, 0x6f, 0x73, 0x74,
+            0x2e, 0x6c, 0x6f, 0x63, 0x61, 0x6c, 0x64, 0x6f, 0x6d, 0x61, 0x69,
+            0x6e, /* Server framebuffer parameters */
+        ];
+        init_state.parse_response(&buf[0..12]);
+        let mut ok_state = parser::RFBGlobalState::TSClientProtocolVersion;
+        assert_eq!(init_state.state, ok_state);
+
+        init_state.parse_request(&buf[0..12]);
+        ok_state = parser::RFBGlobalState::TCSupportedSecurityTypes;
+        assert_eq!(init_state.state, ok_state);
```

Comment:
3) In situations where you use the same buffer as input for `parse_response` and `parse_request`, add a comment explaining why, as it's not so expected that the same buffer can act as a response and as a request.

Reply:
In real-time, they are going to be two separate streams of values. I used the same one here because both corresponded to the protocol version. Is that okay?

Reply:
Can you elaborate a bit more? I'm not sure I understand what you mean

Reply:
In real-time, there will be two separate buffers e.g `buff1 = [0x52, 0x46, 0x42, 0x20, 0x30, 0x30, 0x33, 0x2e, 0x30, 0x30, 0x38, 0x0a] and buff2 = [0x52, 0x46, 0x42, 0x20, 0x30, 0x30, 0x33, 0x2e, 0x30, 0x30, 0x38, 0x0a]`
 Buff
Here I am passing the same because of similar values.

Reply:
I see. So does this mean that this portion of the message header is the same for both request and response? If some, I guess that's alright - so long as you add a comment explaining this.

Is it the same thing that is happening from `15..31`?

Reply:
yes

Reply:
Then, same answer :P